### ISSUE TITLE
How i can change zIndex of elements?

### ISSUE THREAD

**Issue by lmacsen** [pixijs/pixi.js#300]
Hello guys,

I have two shapes rendered on my canvas, on click I want to change zindex of element and bring it to front.

Is there a method that can change zindex of stage children? Or I need to clear and render again the content of container?

![screen shot 2013-08-31 at 10 52 17 am](https://f.cloud.github.com/assets/5254808/1061916/5e41aa60-1213-11e3-9f48-93b700e29e86.png)
Thank you!

**Comment by tengotengo** [comment]
```javascript
var mapContainer = new PIXI.DisplayObjectContainer(),
    unitsContainer = new PIXI.DisplayObjectContainer(),
    menuContainer = new PIXI.DisplayObjectContainer();

mapContainer.zIndex = 5;
unitsContainer.zIndex = 10;
menuContainer.zIndex = 20;

/* adding children, no matter in which order */
stage.addChild(mapContainer);
stage.addChild(menuContainer);
stage.addChild(unitsContainer);

/* call this function whenever you added a new layer/container */
stage.updateLayersOrder = function () {
    stage.children.sort(function(a,b) {
        a.zIndex = a.zIndex || 0;
        b.zIndex = b.zIndex || 0;
        return a.zIndex - b.zIndex
    });
};

stage.updateLayersOrder();

```

**Comment by lmacsen** [comment]
Thank @tengotengo, nice solution.

**Comment by littlewords** [comment]
Than @tengotengo , get it

**Comment by loadedsith** [comment]
If you're here looking for this feature in 2017; it's been implemented in Pixi as zOrder.

https://pixijs.github.io/examples/#/display/zorder.js

**Comment by NavCore** [comment]
Thanks @loadedsith.

Does the zOrder works for each sprite or just for the sprites in DisplayGroup?

**Comment by ivanpopelyshev** [comment]
For sprites in DisplayGroup. Its special optimization: suppose you have 1002 sprites and you want to bring only 2 of them to front. Then sorting will work only for those 2, its much faster than for all 1002.

Also, there's new iteration of plugin: https://pixijs.github.io/examples/#/layers/lighting.js , its called pixi-layers

**Comment by NavCore** [comment]
Thanks @ivanpopelyshev.

And if I have for example 10 sprites and I need to remove and add them multiple times but each sprite needs to be zOrdered. How to accomplish that?

I'm already using stage.addChildAt() method. 

Can add each of them in DisplayGroup and zOrder DisplayGroup or?

Thanks

**Comment by ivanpopelyshev** [comment]
if you are using old "pixi-display":

1. create a PIXI.DisplayGroup with sorting enabled
2. assign it to a displayGroup field of sprites you want to sort
3. create a PIXI.DisplayList
4. assign it to some root container in "displayList" field

Every frame it'll sort them again, without changing their actual order in "children".

**Comment by ivanpopelyshev** [comment]
If you are using new "pixi-layers":

1. create PIXI.display.Group with sorting enabled.
2. specify it in "parentGroup" field of sprites you want to sort
3. create a PIXI.display.Layer and add it as a child somewhere
4. specify "layer.group = myGroup"

Now, every frame all sprites with same displayGroup will be sorted and then rendered inside that Layer.

**Comment by kaansoral** [comment]
Can you give more information about "pixi-layers", is it yours? is it better? is there a github page?

Thanks :)

**Comment by ivanpopelyshev** [comment]
right now its just a branch in pixi-display. Its new iteration that addresses issue of moving objects inside containers with filters & masks. 

https://github.com/pixijs/pixi-display/tree/layers

The downside is that you have to create layer containers where elements with corresponding "parentGroup" or "parentLayer" will be rendered. The good side is that its optimal for lighting :) I'm gonna add examples with normal maps and other advanced lighting.

**Comment by NavCore** [comment]
Great! Can You give some additional usage tips?

As I can see, we create a layer that extends container:
`var layer = new PIXI.display.Layer();`

Then we can add our sprites in that layer:
`bunnySprite.parentLayer = layer;`

"Layer can order elements inside of it, by zIndex increase and then by zOrder decrease."
```
bunnySprite.zIndex = 1;
cloudSprite.zIndex = 2;
badCloudSprite.zIndex = 2;
badCloudSprite.zOrder = 1;
layer.group.enableSort = true;
```

Can You explain this zIndex and zOrder a little better?

I need one layer to add all my sprites in it and then specify zOrder (if I'm right) - so that renderer knows which sprite to render over the other?

And what about PIXI.Container? Currently all my sprites are inside PIXI.Container.
`var stage = new PIXI.container;`

Just to change it to:
`var stage = new PIXI.display.Layer();`

.. and add sprites to stage using:
`mySprite.parentLayer  = stage;`

Thanks in advanced!

**Comment by ivanpopelyshev** [comment]
The idea is that you can combine your objects in containers whatever you like. Imagine that you are combining character, its nickname and its shadow in the same container, because its easy to move it, add it or delete it. Shadows go to one layer, character to another, and nickname to the third.

```js
app.stage = new PIXI.display.Stage();

var myObjects = new PIXI.Container();
stage.addChild(myObjects);

var sortedLayer = new PIXI.display.Layer();
sortedLayer.group.enableSort = true;
stage.addChild(sortedLayer);

var someObj = createMyObject();
myObjects.addChild(someObj);
someObj.parentLayer = sortedLayer;

```

**Comment by ivanpopelyshev** [comment]
And you need to create PIXI.display.Stage, otherwise nothing will work.

**Comment by ivanpopelyshev** [comment]
That's how it works:

1. Stage looks into all elements, and adds them to corresponding layers as an "active children", its re-calculated every time.
2. When myObjects asks his child to render, the child goes "oh, wait, wrong time, I'll do it later"
3. When sortedLayer asks his "active children" to render, they draw themselves.

**Comment by ivanpopelyshev** [comment]
OK, I've got this, better analogy:

actual pixi children, addChild, removeChild - that's all are Homes/flats

layers are offices. Office for coders, office for electrics, and such. 

groups are occupation. Office is for people with certain occupation.

**Comment by NavCore** [comment]
@ivanpopelyshev -> mate You done a great job, I got it !! 
** Your tips are very usefull =)

_It's already implement in my code ;)_

**Thanks**!

**Comment by starogre** [comment]
@ivanpopelyshev do i need to update pixi to some latest iteration to be able to use layers you describe? it's kind of confusing but i'm willing to figure it out. new to this. i was trying to come up with my own solution earlier, but at the moment, just trying to extract the layers of top-down tilebased terrain out so i can fit player in between them.

**Comment by ivanpopelyshev** [comment]
You must have PIXI.display.Stage as root

```
var stage = new PIXI.display.Stage();
var container = new PIXI.Container();
var topLayer = new PIXI.display.Layer();

stage.addChild(container);
stage.addChild(topLayer);

for (var i=0;i<10;i++) { container.addChild(new Sprite(myTexture)) }

container.children[7].parentLayer = topLayer
// 7-th child now rendered in topLayer instead of container, and its above everything else

renderer.render(stage);
```

**Comment by starogre** [comment]
Awesome thanks! Any advice on how to install your branch etc?

**Comment by ivanpopelyshev** [comment]
Just add this js file: https://github.com/ivanpopelyshev/pixi-legacy/tree/master/bin

**Comment by starogre** [comment]
Cool thanks for the help. One last question, what do you mean by removing pixi interactive from working with it? Will I still be able to do character collision or does it mean that something like a light filter you're using in your example just won't have collision for instance

**Comment by ivanpopelyshev** [comment]
Yeah, the problem is if character is interactive, and someone clicks on light sprite, event will bubble to character, you have to be aware of that, and amy be prevent it somehow.

Also, that plugin hacks interaction algorithm, it can be a bit buggy. However,  there were no issues with it so far :)

**Comment by ivanpopelyshev** [comment]
May be i'll have to add one more feature to groups: make all sprites of that group non-interactable

**Comment by starogre** [comment]
Alright cool. I'll have to look into using groups with Display. Right now I'm just rendering everything to a pixi container.

**Comment by Zzzen** [comment]
Why not implement z-order? `pixi-layers` is just a temp workaround. This issue should be reopened. @ivanpopelyshev

**Comment by ivanpopelyshev** [comment]
I hope to add pixi-layers in new pixi, v5. 

The idea is that layers give much more functionality than just z-order, I dont know of any rendering libs that can do sorting through the tree. 

As for simple solutions, https://github.com/pixijs/pixi-display/wiki - may be we can add SortableContainer to pixi?

**Comment by bQvle** [comment]
Yeah it's very handy. I don't see any good way of doing proper draw call optimization with 'pure' pixi. We need z-index or layers to do that. let's say you have a player container, you put everything into that container. including the name text (font atlas). you end up with "roughly" 3 draw calls. now add 10 players. since it'll render one player at a time you will end up with about 30 draw calls, switching between your spritesheet and font atlas.

With pixi layers you could add 3 dedicated layers for your player rendering and do something like.

```
PlayerName.parentLayer = NameLayer;
PlayerHealth.parentLayer = HealthLayer;
PlayerBody.parentLayer = PlayerLayer;
```

So now we are back to 3 draw calls.
1. render all bodies
2. render all healthbars
3. render all names

Besides being great for optimization, layers adds some interesting possibilities as well.
We have layers for normal particles, particles with blendmode "add" etc. so most of our particles happens in two draw calls. We also made a special layer for displacement. so adding `parentLayer = displacementLayer` would instantly make our sprite displace the stage. and we can use alpha to adjust the displacement amount.

Check out this predator effect. all it does is render a "clone" of the tank in our displacement layer. then crossfade the alpha between the two tanks.

![predator](https://user-images.githubusercontent.com/12952595/32996016-5852ebe4-cd7d-11e7-8947-e4a6bffc3789.gif)

We are so heavily invested in pixi-layers that we wouldn't be able to migrate to v5 until layers work. Hopefully @GoodBoyDigital / @englercj (and everyone who has a vote in v5) will have a serious look at layers. It's really a great tool, and a missing feature.

**Comment by NathanHawks** [comment]
This feature doesn't appear to be in the manual yet.* Is it available in 4.8.2?

* unless i'm using the wrong manual: http://pixijs.download/release/docs/PIXI.DisplayObject.html#renderCanvas
